File: eRents.Domain/Models/LeaseExtensionRequest.cs
using System;
using eRents.Domain.Shared;
using eRents.Domain.Models.Enums;

namespace eRents.Domain.Models;

public class LeaseExtensionRequest : BaseEntity
{
    public int LeaseExtensionRequestId { get; set; }
    public int BookingId { get; set; }
    public int RequestedByUserId { get; set; }

    public DateOnly? OldEndDate { get; set; }
    public DateOnly? NewEndDate { get; set; }
    public int? ExtendByMonths { get; set; }
    public decimal? NewMonthlyAmount { get; set; }

    public string? Reason { get; set; }
    public LeaseExtensionStatusEnum Status { get; set; } = LeaseExtensionStatusEnum.Pending;

    public int? RespondedByUserId { get; set; }
    public DateTime? RespondedAt { get; set; }

    public virtual Booking Booking { get; set; } = null!;
}

File: e_rents_desktop/lib/models/lease_extension_request.dart
class LeaseExtensionRequest {
  final int requestId;
  final String status;
  final DateTime createdAt;
  final DateTime? respondedAt;
  final int requestedByUserId;
  final String? requestedByUserName;
  final int bookingId;
  final int propertyId;
  final String propertyName;
  final DateTime? oldEndDate;
  final DateTime? newEndDate;
  final int? extendByMonths;
  final double? newMonthlyAmount;

  LeaseExtensionRequest({
    required this.requestId,
    required this.status,
    required this.createdAt,
    this.respondedAt,
    required this.requestedByUserId,
    this.requestedByUserName,
    required this.bookingId,
    required this.propertyId,
    required this.propertyName,
    this.oldEndDate,
    this.newEndDate,
    this.extendByMonths,
    this.newMonthlyAmount,
  });

  /// Display name for the requester - shows name if available, otherwise user ID
  String get requesterDisplayName => 
      requestedByUserName?.isNotEmpty == true ? requestedByUserName! : 'User #$requestedByUserId';

  factory LeaseExtensionRequest.fromJson(Map<String, dynamic> json) => LeaseExtensionRequest(
        requestId: json['leaseExtensionRequestId'] as int,
        status: (json['status'] as String?) ?? 'Pending',
        createdAt: DateTime.parse(json['createdAt'] as String),
        respondedAt: json['respondedAt'] != null ? DateTime.parse(json['respondedAt'] as String) : null,
        requestedByUserId: json['requestedByUserId'] as int,
        requestedByUserName: json['requestedByUserName'] as String?,
        bookingId: json['bookingId'] as int,
        propertyId: json['propertyId'] as int,
        propertyName: json['propertyName'] as String,
        oldEndDate: _parseDate(json['oldEndDate']),
        newEndDate: _parseDate(json['newEndDate']),
        extendByMonths: json['extendByMonths'] as int?,
        newMonthlyAmount: (json['newMonthlyAmount'] as num?)?.toDouble(),
      );

  static DateTime? _parseDate(dynamic v) {
    if (v == null) return null;
    final s = v.toString();
    // Support DateOnly serialized as yyyy-MM-dd
    if (s.length == 10 && s[4] == '-' && s[7] == '-') {
      return DateTime.parse(s);
    }
    return DateTime.tryParse(s);
  }
}
